Implements job skillchip framework as per hackmd.io design document

Timberpoes opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Implements https://hackmd.io/CvKcHVxCTrKEEAdXJ8ntyA

Adds/re-adds skill stations to Delta and Pubby.

### Implements framework for job skillchips.

Job skillchips are mutually exclusive with each other; implanting a new chip requires removing an old chip.

They have a cooldown which can be polled to decide if they are safe to remove. Allows for example for a surgery to rip out a skill chip pre-emptively with risks. The default Skill Station will not remove a chip if it is unsafe to do so.

Skill Station is modified to support these changes.

Job skillchips are defined in the job outfit datum.

A box containing the new job skill chips have been added to the relevant head lockers.

### Implements first two job chips

Roboticist skill chip - Allows the implanted person to see cyborg wire assignments as if they were holding a blueprint.

Station engineer skill chip - Allows them to see airlock and APC wire assignments as if they were holding a blueprint.

### Implements code paths for two major antag types interacting with skill chips in general.

Traitors now gain access to a chameleon skill chip. This chip costs 4TC and comes with a single-use auto-surgeon that instantly implants it. It can be used once every 5 minutes and acts as any other chameleon item, allowing the user to select from any skillchip - Job or otherwise. Fully compatible with chameleon outfit selection.

Changelings, on the other hand, replicate the skillchips of their targets but lose their own in the process. They retain their last set of skillchips when they enter lesser form.

None of these skill chips are recoverable. On removal, they are destroyed.

### Code updates

Refactors part of wire datum code, moving hardcoded logic to an overridable proc to allow wires to be revealed on a case-by-case basis.

Subtypes autosurgeons to have both organ/implant and skillchip variants.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Job skillchips are a fun new replacement for grindable skills. They're modular, modifiable and easy to use and implement.

You can only ever have one implanted at once and have to remove an old one to use a new one.

They can provide bonuses to jobs, yet can also be stolen. Either from head lockers or directly from the brains of those who have them.

It's up to you to think of the fun and creative ways you're going to go about obtaining the job chips you feel you need. From TC to theft, to HARS and human harm, you too can trascend your former job title and become something greater.

Implements the hackmd.io design document.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Roboticists and engineers rejoice. NanoTrasen has decided to cut back on training costs through leveraging emerging technologies. All roboticists and engineers start with tiny little microchips implanted deep in their brains that unlock forbidden knowledge about wire assignments.
add: Roboticists with their job skillchip can now identify cyborg wires while engineers with their job skillchip can now identify airlock and APC wires. Highly select blueprint-based knowledge implanted directly into your brain.
add: The Research Director and Chief Engineer now have a box containing a couple of spare copies of their respective department's skill chips tucked nearly away in their secure lockers. Feel free to implant yourselves - As the most competent and highly respected members of your respective departments, we're sure you're earned it...
add: Traitors not left out of the loop either, they can now purchase a highly advanced Syndicate skillchip for 4TC that contains data a number of key skillchips. This so-called "chameleon" chip will enable you to inflitrate other departments or take full advantage of the powers the skillchips have to offer; just remember to extract your own job skillchips first at any library before going undercover if you want your own chameleon chip to work.
add: Changelings rejoice, for you now absorb the skill chips of your targets when you transform. Beware, though - You'll lose all your existing ones in the process. A truly double-edged armblade.
add: Delta and Pubby now have Skill Stations in their respective libraries, next to the games vending machines.
tweak: Skill Station has recieved a fresh coat of interface paint.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Timberpoes committed:
Shift wire revealing logic to dedicated proc

Timberpoes committed:
Bit of code modification. Framework for roundstart job skillchips.

Timberpoes committed:
Implement roboticist skillchip trait functionality.

Timberpoes committed:
Feex moth feet messup

Timberpoes committed:
Add skill chips to robodrobe inventory

Timberpoes committed:
Code refactor. Add missing skill_stations to Pubby and Delta.

Timberpoes committed:
Implement special flags, changelings inherit skillchip skills

Timberpoes committed:
Additional code refactor. Traitor chameleon skillchip framework.

Timberpoes committed:
Implement traitor skillchip, fix up more code, move job chips to outfit datums

Timberpoes committed:
Modify autosurgeon, create skillchip variant, add to uplink

Timberpoes committed:
Implement chip cooldowns. Add new skillchip for station engineers. Cleanup some code. Add job chips to lockers.

Timberpoes committed:
Feex

Timberpoes committed:
Feex

Fikou commented on code/modules/uplink/uplink_items.dm:
```diff
@@ -1185,11 +1185,20 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
+			Comes with a single-use Syndicate autosurgeon for immediate self-application."
+	item = /obj/item/autosurgeon/skillchip/syndicate/chameleon_chip
+	cost = 4
+	exclude_modes = list(/datum/game_mode/nuclear)
```
not gonna include clown ops?

Timberpoes replied:
Nope. None of the other chameleon gear excludes clown ops. I figured I'd follow the theme.

Time-Green commented:
> None of these skill chips are recoverable. On removal, they are destroyed.

> They can provide bonuses to jobs, yet can also be stolen. Either from head lockers or directly from the brains of those who have them.

Wait, which one is it? Personally prefer them being stealable. 

Looks great, finally a way to have job-skills that arent hardcoded or require an hour to have any effect.

Timberpoes commented:
> 
> 
> > None of these skill chips are recoverable. On removal, they are destroyed.
> 
> > They can provide bonuses to jobs, yet can also be stolen. Either from head lockers or directly from the brains of those who have them.
> 
> Wait, which one is it? Personally prefer them being stealable.
> 
> Looks great, finally a way to have job-skills that arent hardcoded or require an hour to have any effect.

Both. Specifically the changeling's faux-chips they create as part of transforming into a person are non-recoverable. Similarly, the traitor's chameleon chip is non-recoverable. Meanwhile, ordinary chips can be yanked out of the brains of poor, poor saps.

tralezab commented:
Changelings should not be able to reproduce mechanical chips, only biological parts.

Rohesie commented on code/__DEFINES/flags.dm:
```diff
@@ -157,3 +157,9 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define SKILLCHIP_ALLOWS_MULTIPLE (1<<0)
+// Job skillchip, only one job skillchip can be implanted.
+#define SKILLCHIP_JOB_TYPE (1<<1)
+// Skillchip requires the target to be mindshielded to be implanted.
+#define SKILLCHIP_REQUIRE_MINDSHIELD (1<<2)
```
Maybe something like `SKILLCHIP_EXTRA_REQUIREMENTS` for a re-definable check such as  `if((skillchip.skillchip_flags & SKILLCHIP_EXTRA_REQUIREMENTS) && !implant_checks(src))` ?
That would be more expansible.

Rohesie commented on code/__DEFINES/flags.dm:
```diff
@@ -157,3 +157,9 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 //skillchip flags
 //Skillchip type can be implanted multiple times
 #define SKILLCHIP_ALLOWS_MULTIPLE (1<<0)
+// Job skillchip, only one job skillchip can be implanted.
+#define SKILLCHIP_JOB_TYPE (1<<1)
```
`SKILLCHIP_HIGHLANDER` or something to describe less the specificity of this condition, and more the general property?

Rohesie commented on code/__DEFINES/flags.dm:
```diff
@@ -157,3 +157,9 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
+// Skillchip requires the target to be mindshielded to be implanted.
+#define SKILLCHIP_REQUIRE_MINDSHIELD (1<<2)
+// Used by the traitor skillchip, not implantable unless it's set to a copy of another skillchip.
+#define SKILLCHIP_CHAMELEON (1<<3)
```
Same here, separate the generic conditions that make the traitor skillchip unique and add them here, instead of bundling up into a single flag.

Rohesie commented on code/game/gamemodes/changeling/changeling.dm:
```diff
@@ -110,6 +110,18 @@ GLOBAL_VAR(changeling_team_objective_type) //If this is not null, we hand our th
+
+	// Add new set of skillchips.
+	for(var/S in chosen_prof.skillchips)
+		var/obj/item/skillchip/skillchip = new S()
+		skillchip.removable = FALSE
```
Could you add the `removable` change to the chip's initialize, as an argument?

Rohesie commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -1360,3 +1360,29 @@
+/obj/item/storage/box/skillchips/PopulateContents()
+	var/list/skillchips = subtypesof(/obj/item/skillchip)
+
+	for(var/S in skillchips)
+		new S(src)
```
```suggestion
	for(var/skill_chip in skillchips)
		new skill_chip(src)
```

Rohesie replied:
Still a thing.

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +681,47 @@
+/datum/action/item_action/chameleon/change/skillchip
+	/// Skillchip this this chameleon action is imitating.
+	var/obj/item/skillchip/skillchip_mimic
+	var/obj/item/skillchip/usable_at_world_time = 0
```
```suggestion
	COOLDOWN_DECLARE(usable_at_world_time)
```

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +681,47 @@
+		target.icon_state = initial(picked_item.skill_icon)
+
+/datum/action/item_action/chameleon/change/skillchip/update_look(mob/user, obj/item/skillchip/picked_item)
+	if(usable_at_world_time <= world.time)
+		// Swap out the fake skillchips.
+		if(skillchip_mimic && istype(skillchip_mimic))
+			skillchip_mimic.on_removal(user, silent = FALSE)
+			QDEL_NULL(skillchip_mimic)
+
+		skillchip_mimic = new picked_item()
+		if(istype(skillchip_mimic))
+			skillchip_mimic.removable = FALSE
+			skillchip_mimic.on_apply(user, silent = FALSE)
+			usable_at_world_time = world.time + cooldown
+			to_chat(user, "<span class='notice'>The chameleon skillchip is recharging. It will be unable to change for another [cooldown/10] seconds.</span>")
+		else
+			QDEL_NULL(skillchip_mimic)
+	else
+		to_chat(user, "<span class='notice'>Chameleon skillchip is still recharging for another [(usable_at_world_time - world.time)/10] seconds!</span>")
+
+	..()
```
```suggestion
	if(!COOLDOWN_FINISHED(src, usable_at_world_time))
		to_chat(user, "<span class='notice'>Chameleon skillchip is still recharging for another [COOLDOWN_TIMELEFT(src, usable_at_world_time) * 0.1] seconds!</span>")
		return ..()

	// Swap out the fake skillchips.
	if(istype(skillchip_mimic))
		skillchip_mimic.on_removal(user, silent = FALSE)
		QDEL_NULL(skillchip_mimic)

	skillchip_mimic = new picked_item()
	if(!istype(skillchip_mimic))
		QDEL_NULL(skillchip_mimic)
		return ..()

	COOLDOWN_START(src, usable_at_world_time, cooldown)
	skillchip_mimic.removable = FALSE
	skillchip_mimic.on_apply(user, silent = FALSE)
	to_chat(user, "<span class='notice'>The chameleon skillchip is recharging. It will be unable to change for another [cooldown * 0.1] seconds.</span>")
	return ..()
```
* Early returns.
* Cooldown macros.
* istype already returns a falsey value for a null/falsey input.
* Multiplications over divisions whenever possible.

Rohesie commented on code/modules/jobs/job_types/_job.dm:
```diff
@@ -260,12 +262,20 @@
 
+	// Insert any skillchips associated with this job into the target's brain.
+	if(skillchip && istype(H))
+		var/obj/item/skillchip/S = new skillchip()
+		if(!H.implant_skillchip(S))
+			qdel(S)
```
```suggestion
		var/obj/item/skillchip/skill_chip = new skillchip()
		if(!H.implant_skillchip(skill_chip))
			stack_trace("Failed to implant [H] with [skill_chip], on job [src].")
			qdel(skill_chip)
```

Rohesie commented on code/modules/library/skill_learning/job_skillchips/roboticist.dm:
```diff
@@ -0,0 +1,13 @@
+	implanting_message = "<span class='notice'>You suddenly comprehend the secrets behind cyborg circuitry.</span>"
+	removal_message = "<span class='notice'>Cyborg circuitry stops making sense as images of coloured wires fade from your mind.</span>"
+
+/obj/item/skillchip/roboticist/Initialize()
+	. = ..()
+	skillchip_flags |= SKILLCHIP_JOB_TYPE
```
You can add this to the object definition, it doesn't need to be on `Initialize()`
Just know beforehand which are the appropriate bitflags.

Rohesie commented on code/modules/library/skill_learning/job_skillchips/station_engineer.dm:
```diff
@@ -0,0 +1,13 @@
+	implanting_message = "<span class='notice'>You suddenly comprehend the secrets behind airlock and APC circuitry.</span>"
+	removal_message = "<span class='notice'>Airlock and APC circuitry stops making sense as images of coloured wires fade from your mind.</span>"
+
+/obj/item/skillchip/roboticist/Initialize()
+	. = ..()
+	skillchip_flags |= SKILLCHIP_JOB_TYPE
```
Same for this one, shouldn't be on `Initialize()`
You might even want to make a subtype for job skillchips.

Rohesie commented on code/modules/library/skill_learning/job_skillchips/traitor.dm:
```diff
@@ -0,0 +1,30 @@
+/obj/item/skillchip/chameleon/on_apply(mob/living/carbon/user, silent = TRUE)
+	. = ..()
+
+	chameleon_action = new(src)
+	chameleon_action.chameleon_type = /obj/item/skillchip
+	chameleon_action.chameleon_name = "Skillchip"
```
Could you add these changes to the action's Init as arguments?

Rohesie commented on code/modules/library/skill_learning/job_skillchips/traitor.dm:
```diff
@@ -0,0 +1,30 @@
+	chameleon_action = new(src)
+	chameleon_action.chameleon_type = /obj/item/skillchip
+	chameleon_action.chameleon_name = "Skillchip"
+	chameleon_action.chameleon_blacklist = typecacheof(list(/obj/item/skillchip, /obj/item/skillchip/chameleon, /obj/item/skillchip/useless_adapter), only_root_path = TRUE)
```
`typecacheof` already handles subtypes, you would achieve the same result with `typecacheof(/obj/item/skillchip)`
That said, do you really need a typecache here?
It would be better to make a variable or flag instead, and check on it if you want to populate a whitelist.

Timberpoes replied:
I'm going to implement a middleground solution and add a flag check alongside this method.

The problem is - I want to exclude certain type paths from the Chameleon item list such as the parent type /obj/item/skillchip.

I can simply add a flag and stop using this blacklist, but then every single child of /obj/item/skillchip HAS to override the flags.

The reason /obj/item/skillchip would have the chameleon incompatibility flag set is not because it should be the default behaviour of skill chips, but because /obj/item/skillchip isn't a REAL skillchip.

Anyone who forgot to clear the flags when inheriting /obj/item/skillchip would have unexpected functionality in their chips. It's effectively an open season for introducing bugs. I suspect this is part of why chameleon items use the blacklist method that they do.

As a result, I'm doing it both ways. I'm going to hardcode "abstract" item paths in chameleon_blacklist however when building the item list, it also checks for the presence of the appropriate flag to exclude the item too.

The end system will be as follows: Real skillchips that want to be chameleon blacklisted can use the flag. Abstract skillchips that exist just as parents and should never actually appear in-game can have their paths added to the blacklist without forcing ALL their children to override thier flags.

Timberpoes replied:
Testing out an alternative system after talking with AnturK on Discord. Could have found a better way,

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -133,24 +141,26 @@
-	if(to_be_removed.removable)
-		carbon_occupant.put_in_hands(to_be_removed)
+
+	if(istype(carbon_occupant))
+		if(carbon_occupant.remove_skillchip(to_be_removed))
+			if(to_be_removed.removable)
+				carbon_occupant.put_in_hands(to_be_removed)
+			else
+				qdel(to_be_removed)
+
+			to_chat(carbon_occupant, "<span class='notice'>Operation complete!</span>")
+		else
+			to_chat(carbon_occupant,"<span class='notice'>Failed to remove skillchip!</span>")
```
Repeat `update_icon` and `SStgui.update_uis(src)` if needed, but turn these into early returns, please.

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -162,16 +172,24 @@
 		var/list/current_skills = list()
-		for(var/obj/item/skillchip/skill_chip in occupant_brain)
-			current_skills += list(list("name"=skill_chip.skill_name,"icon"=skill_chip.skill_icon,"cost"=skill_chip.slot_cost,"ref"=REF(skill_chip),"active"=(skill_chip in occupant_brain.skillchips)))
+		for(var/obj/item/skillchip/skill_chip in occupant_brain.skillchips)
+			current_skills += list(
```
```suggestion
		for(var/sc in occupant_brain.skillchips)
			var/obj/item/skillchip/skill_chip = sc
			current_skills += list(
```

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -162,16 +172,24 @@
+					"icon" = skill_chip.skill_icon,
+					"cost" = skill_chip.slot_cost,
+					"ref" = REF(skill_chip),
+					"active" = (skill_chip in occupant_brain.skillchips),
```
Can this check ever be false, given the skillchip was taken from this precise list?

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -162,16 +172,24 @@
+					"cost" = skill_chip.slot_cost,
+					"ref" = REF(skill_chip),
+					"active" = (skill_chip in occupant_brain.skillchips),
+					"cooldown" = skill_chip.extractable_at - world.time,
```
```suggestion
					"cooldown" = COOLDOWN_TIMELEFT(skill_chip, extractable_at),
```

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -25,6 +25,10 @@
+	/// Cooldown before the skillchip can be extracted after it has been implanted.
+	var/cooldown = 5 MINUTES
+	/// The world.time when this skillchip should be extractable.
+	var/extractable_at = 0
```
```suggestion
	COOLDOWN_DECLARE(extractable_at)
```

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -34,6 +38,8 @@
 		ADD_TRAIT(user,auto_trait,SKILLCHIP_TRAIT)
 	user.used_skillchip_slots += slot_cost
 
+	extractable_at = world.time + cooldown
```
```suggestion
	COOLDOWN_START(src, extractable_at, cooldown)
```

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -42,33 +48,49 @@
-	//Only one multiple copies of a type if SKILLCHIP_ALLOWS_MULTIPLE flag is set
-	if(!(skillchip_flags & SKILLCHIP_ALLOWS_MULTIPLE) && (locate(type) in target_brain.skillchips))
-		return FALSE
+	extractable_at = 0
```
```suggestion
	COOLDOWN_RESET(src, extractable_at)
```

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -42,33 +48,49 @@
+  * * skillchip - The skillchip to test for implantability.
+  */
+/obj/item/skillchip/proc/check_incompatibility(obj/item/skillchip/skillchip)
+	var/incompatible_flags = 0
+
+	// If this is a SKILLCHIP_JOB_TYPE it is incompatible with any other SKILLCHIP_JOB_TYPE.
+	if((skillchip_flags & SKILLCHIP_JOB_TYPE) && (skillchip.skillchip_flags & SKILLCHIP_JOB_TYPE))
+		incompatible_flags |= SKILLCHIP_JOB_TYPE
+
+	// Only allow multiple copies of a type if SKILLCHIP_ALLOWS_MULTIPLE flag is set
+	if(!(skillchip_flags & SKILLCHIP_ALLOWS_MULTIPLE) && (istype(skillchip, type)))
+		incompatible_flags |= SKILLCHIP_ALLOWS_MULTIPLE
+
+	return incompatible_flags
```
```suggestion
	. = NONE

	// If this is a SKILLCHIP_JOB_TYPE it is incompatible with any other SKILLCHIP_JOB_TYPE.
	if((skillchip_flags & SKILLCHIP_JOB_TYPE) && (skillchip.skillchip_flags & SKILLCHIP_JOB_TYPE))
		. |= SKILLCHIP_JOB_TYPE

	// Only allow multiple copies of a type if SKILLCHIP_ALLOWS_MULTIPLE flag is set
	if(!(skillchip_flags & SKILLCHIP_ALLOWS_MULTIPLE) && (istype(skillchip, type)))
		. |= SKILLCHIP_ALLOWS_MULTIPLE

```
* Remember to use the `NONE` define for no flags.

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -42,33 +48,49 @@
+  * Returns FALSE if the chip's extraction cooldown hasn't yet passed.
+  */
+/obj/item/skillchip/proc/can_remove_safely()
+	if(extractable_at > world.time)
```
```suggestion
	if(!COOLDOWN_FINISHED(src, extractable_at))
```

Rohesie commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -130,10 +130,13 @@
-				if(skill_chip.removable)
-					skill_chip.forceMove(drop_location())
+				if(remove_skillchip(skill_chip))
+					if(skill_chip.removable)
+						skill_chip.forceMove(drop_location())
+					else
+						qdel(skill_chip)
 				else
-					qdel(skill_chip)
+					CRASH("Skillchip could not be extracted from brain as either chip or doesn't exist or the chip is not in the brain.")
```
* Early continues.
* Remove the typecheck on the loop.

Rohesie commented on code/modules/mob/living/brain/skillchip.dm:
```diff
@@ -0,0 +1,70 @@
+  */
+/obj/item/organ/brain/proc/remove_skillchip(obj/item/skillchip/skillchip)
+	// Check this skillchip is in the brain.
+	if(!(skillchip in skillchips))
+		return FALSE
```
```suggestion
	if(!(skillchip in skillchips))
		CRASH("Attempted to remove skillchip not in the skillchips list.")
```
Let's handle cases where this happens if we can.

Rohesie commented on code/modules/mob/living/brain/skillchip.dm:
```diff
@@ -0,0 +1,70 @@
+  * * skillchip - The skillchip you'd like to compare the flags of.
+  */
+/obj/item/organ/brain/proc/check_skillchip_flags(obj/item/skillchip/skillchip)
+	var/incompatibility_flags = 0
+
+	// If this is a job skillchip, check if any other skillchips are present.
+	for(var/obj/item/skillchip/S in skillchips)
+		incompatibility_flags |= skillchip.check_incompatibility(S)
+
+	return incompatibility_flags
```
```suggestion
	. = NONE

	// If this is a job skillchip, check if any other skillchips are present.
	for(var/skill_chip in skillchips)
. |= skillchip.check_incompatibility(skill_chip)

```
* `NONE` define
* Better variable name
* No need to typecheck.

Rohesie commented on code/modules/mob/living/brain/skillchip.dm:
```diff
@@ -0,0 +1,70 @@
+  * Returns a simple list of typepaths.
+  */
+/obj/item/organ/brain/proc/get_skillchip_type_list()
+	var/list/skillchip_types = list()
+	// Remove and call on_removal proc if successful.
+	for(var/obj/item/skillchip/S in skillchips)
+		skillchip_types += S.type
+
+	return skillchip_types
```
```suggestion
	. = list()
	// Remove and call on_removal proc if successful.
	for(var/s in skillchips)
		var/obj/item/skillchip/skill_chip = s
		. += skill_chip.type

```

Rohesie commented on code/modules/mob/living/brain/skillchip.dm:
```diff
@@ -0,0 +1,70 @@
+  */
+/obj/item/organ/brain/proc/destroy_all_skillchips(silent = TRUE)
+	if(!QDELETED(owner))
+		for(var/obj/item/skillchip/skill_chip in skillchips)
+			skill_chip.on_removal(owner, silent)
```
```suggestion
		for(var/sc in skillchips)
			var/obj/item/skillchip/skill_chip = sc
			skill_chip.on_removal(owner, silent)
```

Rohesie commented on code/modules/mob/living/carbon/skillchip.dm:
```diff
@@ -0,0 +1,154 @@
+  */
+/mob/living/carbon/proc/implant_skillchip(obj/item/skillchip/skillchip, silent = FALSE, force = FALSE)
+	// Check the chip can actually be implanted.
+	if(!can_implant_skillchip(skillchip) && !force)
```
```suggestion
	if(!force && !can_implant_skillchip(skillchip))
```
Prioritize the cheaper checks first.

Rohesie commented on code/modules/mob/living/carbon/skillchip.dm:
```diff
@@ -0,0 +1,154 @@
+  */
+/mob/living/carbon/proc/can_implant_skillchip(obj/item/skillchip/skillchip)
+	// Not a skillchip or skillchip doesn't exist.
+	if(!skillchip || !istype(skillchip))
```
```suggestion
	if(!istype(skillchip))
```

Rohesie commented on code/modules/mob/living/carbon/skillchip.dm:
```diff
@@ -0,0 +1,154 @@
+  */
+/mob/living/carbon/proc/can_implant_skillchip(obj/item/skillchip/skillchip)
+	// Not a skillchip or skillchip doesn't exist.
+	if(!skillchip || !istype(skillchip))
+		return FALSE
```
```suggestion
	if(!istype(skillchip))
		CRASH("can_implant_skillchip called with an invalid skillchip ([skillchip])")
```

Rohesie commented on code/modules/mob/living/carbon/skillchip.dm:
```diff
@@ -0,0 +1,154 @@
+  */
+/mob/living/carbon/proc/can_implant_skillchip_message(obj/item/skillchip/skillchip)
+	// Not a skillchip or skillchip doesn't exist.
+	if(!skillchip || !istype(skillchip))
```
```suggestion
	if(!istype(skillchip))
```

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,128 @@
+	if(starting_skillchip)
+		insert_skillchip(new starting_skillchip(src))
+
+/obj/item/autosurgeon/skillchip/proc/insert_skillchip(var/obj/item/skillchip/I)
+	if(!istype(I))
+		return
+	stored_skillchip = I
+	I.forceMove(src)
```
```suggestion
/obj/item/autosurgeon/skillchip/proc/insert_skillchip(var/obj/item/skillchip/skill_chip)
	if(!istype(skill_chip))
		return
	stored_skillchip = skill_chip
	skill_chip.forceMove(src)
```
* Better var names.

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,128 @@
+		to_chat(user, "<span class='alert'>[stored_skillchip] cannot be implanted. [fail_message]</span>")
+		return
+
+	if(user.implant_skillchip(stored_skillchip))
+		to_chat(user,"<span class='notice'>Operation complete!</span>")
+	else
+		to_chat(user,"<span class='notice'>Operation failed!</span>")
+		return
```
```suggestion
	if(!user.implant_skillchip(stored_skillchip))
		to_chat(user,"<span class='notice'>Operation failed!</span>")
		return

		to_chat(user,"<span class='notice'>Operation complete!</span>")
```
* Early returns.

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,128 @@
+	if(!uses)
+		desc = "[initial(desc)] Looks like it's been used up."
+
+/obj/item/autosurgeon/skillchip/attackby(obj/item/skillchip/I, mob/user, params)
+	if(istype(I, skillchip_type))
+		if(stored_skillchip)
+			to_chat(user, "<span class='alert'>[src] already has a skillchip stored.</span>")
+			return
+		else if(!uses)
+			to_chat(user, "<span class='alert'>[src] has already been used up.</span>")
+			return
+		if(!user.transferItemToLoc(I, src))
+			return
+		stored_skillchip = I
+		to_chat(user, "<span class='notice'>You insert the [I] into [src].</span>")
+	else
+		return ..()
```
```suggestion
/obj/item/autosurgeon/skillchip/attackby(obj/item/I, mob/user, params)
	if(!istype(I, skillchip_type))
		return ..()

	if(stored_skillchip)
		to_chat(user, "<span class='alert'>[src] already has a skillchip stored.</span>")
		return

	if(!uses)
		to_chat(user, "<span class='alert'>[src] has already been used up.</span>")
		return

	if(!user.transferItemToLoc(I, src))
		return

	stored_skillchip = I
	to_chat(user, "<span class='notice'>You insert the [I] into [src].</span>")
```
* Keep the argument's type consistent with their possible values, for clarity.
* Early returns.

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,128 @@
+		return ..()
+
+/obj/item/autosurgeon/skillchip/screwdriver_act(mob/living/user, obj/item/I)
+	if(..())
+		return TRUE
```
```suggestion
	. = ..()
	if(.)
		return
```

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,128 @@
+		return TRUE
+
+	var/atom/drop_loc = user.drop_location()
+	for(var/J in src)
+		var/atom/movable/AM = J
+		AM.forceMove(drop_loc)
```
```suggestion
	for(var/thing in contents)
		var/atom/movable/movable_content = thing
		movable_content.forceMove(drop_loc)
```

AnturK commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +681,47 @@
+	var/cooldown = 5 MINUTES
+
+/datum/action/item_action/chameleon/change/skillchip/initialize_disguises()
+	. = ..()
```
I'd prefer if what skillchips are availible here were made explicit by some flag and/or selected subtypes. Since i don't expect all skillchips to be easily available in the future.

Timberpoes replied:
I'm going to implement a middleground solution and add a flag check alongside this method.

The problem is - I want to exclude certain type paths from the Chameleon item list such as the parent type /obj/item/skillchip.

I can simply add a flag and stop using this blacklist, but then every single child of /obj/item/skillchip HAS to override the flags.

The reason /obj/item/skillchip would have the chameleon incompatibility flag set is not because it should be the default behaviour of skill chips, but because /obj/item/skillchip isn't a REAL skillchip.

Anyone who forgot to clear the flags when inheriting /obj/item/skillchip would have unexpected functionality in their chips. It's effectively an open season for introducing bugs. I suspect this is part of why chameleon items use the blacklist method that they do.

As a result, I'm doing it both ways. I'm going to hardcode "abstract" item paths in chameleon_blacklist however when building the item list, it also checks for the presence of the appropriate flag to exclude the item too.

The end system will be as follows: Real skillchips that want to be chameleon blacklisted can use the flag. Abstract skillchips that exist just as parents and should never actually appear in-game can have their paths added to the blacklist without forcing ALL their children to override thier flags.

ExcessiveUseOfCobblestone replied:
what if you just used subtypesof instead of typesof or whatever you're using to build the list? the base type should be the easiest to remove.

AnturK commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -114,16 +115,23 @@
 	if(!to_be_removed)
 		return
+
+	if(!to_be_removed.can_remove_safely())
```
Could use some feedback here.

Timberpoes commented:
> 
> 
> Changelings should not be able to reproduce mechanical chips, only biological parts.

If they don't reproduce the chips, they don't get the effects.

One of the issues with job-based skillchips or even skillchips in general is that "spy-checking" becomes a thing. You can just randomly, at any point in time, ask someone to do something related to a skillchip you know they should have, to prove they're not a traitor/ling/hallucination etc.

It's a flaw in having an equippable item handle a skill-type system and the scope planned for skillchips in the future as more than just trait carriers.

Timberpoes committed:
Code review implementation

Timberpoes committed:
More feexes, improved chameleon chip code and more.

Timberpoes commented:
Right. A huge sweep of changes to address, I hope, all of the review areas. Although I've probably brought up some new review areas.

I've moved a bunch of functionality around again, refactored parts of code. Particular attention should be paid to the Chameleon Skillchips now, which have been made a LOT more robust. They now respect all the restrictions of skillchips, no longer bypass complexity limits, can't duplicate chips already in the user's brain and more.

They also report the chip they're mimicing to the skill station.

Please do another sweep and check the changes. I'm hoping everything is up to spec.

Rohesie commented on code/game/gamemodes/changeling/changeling.dm:
```diff
@@ -110,6 +110,17 @@ GLOBAL_VAR(changeling_team_objective_type) //If this is not null, we hand our th
+	user.destroy_all_skillchips(silent = FALSE)
+
+	// Add new set of skillchips.
+	for(var/S in chosen_prof.skillchips)
+		var/obj/item/skillchip/skillchip = new S(src, FALSE)
```
```suggestion
	for(var/skill_chip in chosen_prof.skillchips)
		var/obj/item/skillchip/skillchip = new skill_chip(src, FALSE)
```
* Better variable names. Come on, I've done this same review several times already on this PR.

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+	COOLDOWN_DECLARE(usable_cooldown)
+	var/cooldown = 5 MINUTES
+
+/datum/action/item_action/chameleon/change/skillchip/Grant(mob/M)
+	if(skillchip_mimic)
+		skillchip_mimic.on_apply(M, silent = FALSE)
+
+	if(!COOLDOWN_FINISHED(src, usable_cooldown))
+		START_PROCESSING(SSfastprocess, src)
+
+	..()
```
```suggestion
/datum/action/item_action/chameleon/change/skillchip/Grant(mob/M)
	. = ..()
	if(skillchip_mimic)
		skillchip_mimic.on_apply(M, silent = FALSE)

	if(!COOLDOWN_FINISHED(src, usable_cooldown))
		START_PROCESSING(SSfastprocess, src)

```

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+
+	STOP_PROCESSING(SSfastprocess, src)
+
+	..()
```
```suggestion
	return ..()

```

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+		button.name = "Change [chameleon_name] Function"
+
+	if(!ispath(chameleon_type, /obj/item/skillchip))
+		stack_trace("Attempted to initialise [src] disguise list with incompatible item path [chameleon_type].")
+		return
```
```suggestion
		CRASH("Attempted to initialise [src] disguise list with incompatible item path [chameleon_type].")
```

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+		return
+
+	for(var/chip_type in typesof(chameleon_type))
+		var/obj/item/skillchip/SC = chip_type
+		if((chip_type == initial(SC.abstract_parent_type)) || (initial(SC.skillchip_flags) & SKILLCHIP_CHAMELEON_INCOMPATIBLE) || (initial(SC.item_flags) & ABSTRACT) || !initial(SC.icon_state))
+			continue
+		var/chameleon_item_name = "[initial(SC.name)] ([initial(SC.icon_state)])"
+		chameleon_list[chameleon_item_name] = SC
```
Better variable names.

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+	var/obj/item/skillchip/new_chip = new picked_item(src, FALSE)
+
+	// Do a bit of a sanity check.
+	if(!istype(new_chip))
+		stack_trace("Chameleon skillchip [src] attempted to change into non-skillchip item [picked_item].")
+		QDEL_NULL(new_chip)
+
```
```suggestion
	if(!istype(new_chip))
		stack_trace("Chameleon skillchip [src] attempted to change into non-skillchip item [picked_item].")
		QDEL_NULL(new_chip)
		return

```

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -681,3 +685,98 @@
+	if(COOLDOWN_FINISHED(src, usable_cooldown))
+		button.maptext = ""
+		UpdateButtonIcon()
+		STOP_PROCESSING(SSfastprocess, src)
+	else
+		button.maptext = "<b>[COOLDOWN_TIMELEFT(src, usable_cooldown) * 0.1]</b>"
```
```suggestion
		STOP_PROCESSING(SSfastprocess, src)
		return
	button.maptext = "<b>[COOLDOWN_TIMELEFT(src, usable_cooldown) * 0.1]</b>"
```
* Early return.

Rohesie commented on code/modules/library/skill_learning/job_skillchips/traitor.dm:
```diff
@@ -0,0 +1,72 @@
+	chameleon_action.Remove(user)
+	user.max_skillchip_slots--
+	user.used_skillchip_slots--
+	. = ..()
+
```
```suggestion
	return ..()


```
* Use `return ..()` at the end of the proc, instead of `. = ..()`

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -104,8 +104,12 @@
+	var/mob/living/carbon/carbon_occupant = occupant
+
+	if(inserted_skillchip.has_mob_incompatibility(carbon_occupant))
+		stack_trace("Unusual error - [usr] attempted to start implanting of [inserted_skillchip] when the interface state should not have allowed it.")
```
`stack_trace` is just `CRASH` without a return.
Use the latter if there is one.

Rohesie commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -114,16 +118,25 @@
+
+	if(!to_be_removed.can_remove_safely())
+		to_chat(occupant, "<span class='notice'>DANGER! Operation cannot be completed, removal is unsafe.</span>")
+		stack_trace("Unusual error - [usr] attempted to start removal of [to_be_removed] when the interface state should not have allowed it.")
+		return
```
```suggestion
		CRASH("Unusual error - [usr] attempted to start removal of [to_be_removed] when the interface state should not have allowed it.")
```

Rohesie commented on code/modules/library/skill_learning/skillchip.dm:
```diff
@@ -42,33 +61,109 @@
+	var/chip_message
+
+	// Check if this chip is incompatible with any other chips in the brain.
+	for(var/obj/item/skillchip/skillchip in brain.skillchips)
+		chip_message = has_skillchip_incompatibility(skillchip)
```
```suggestion
	for(var/skillchip in brain.skillchips)
		chip_message = has_skillchip_incompatibility(skillchip)
```
No need to typecast.

AnturK commented on code/modules/library/skill_learning/skill_station.dm:
```diff
@@ -132,51 +145,84 @@
+
+	// If we got here, we have both an occupant and it has a brain, so we can check for skillchips.
+	var/list/current_skills = list()
+	for(var/obj/item/skillchip/skill_chip in occupant_brain.skillchips)
```
Previously this was all skillchips located in the brain because you want to see/remove inactive skillchips.

Timberpoes replied:
All skillchips should be located in brain.skillchips now, regardless of whether they are active or not.

There's been code refactoring around this area.

Skillchip activity is now denoted by a simple TRUE/FALSE flag that is set when the skillchip is activated or deactivated. In essence, it tells you if the last operation called was on_activate or on_deactivate and as such tells you whether the body has the effects of the skillchip or not.

Activating and de-activating are now events seperate to implating and removing the skillchip, as a result.

There should be no need to check outside of brain.skillchips - All skillchips currently jammed into the brain should be in that list.

Naturally /mob/living/carbon/proc/update_skillchips() is updated as a result. And has had more descriptive var names added so Rohesie doesn't lynch me.

Expect to see that with the review PR, which will likely trigger another review due to additional code changes. Although I'm hoping I've squashed all the stuff Rohesie keeps yelling at me for.

Rohesie commented on code/modules/surgery/organs/autosurgeon.dm:
```diff
@@ -81,24 +89,136 @@
+	// The below check should never return FALSE, but we're checking it for sanity.
+	// If stored_skillchip.has_mob_incompatibility(user) returns FALSE, there should be abosolutely no reason for this to fail, ever.
+	if(!user.implant_skillchip(stored_skillchip))
+		stack_trace("Autosurgeon failed to implant [stored_skillchip] into [user] for an unknown reason.")
+		to_chat(user,"<span class='notice'>Operation failed!</span>")
+		return
```
```suggestion
		to_chat(user,"<span class='notice'>Operation failed!</span>")
		CRASH("Autosurgeon failed to implant [stored_skillchip] into [user] for an unknown reason.")
```

Timberpoes committed:
Code review and minor refactor

Timberpoes committed:
Additional review fixes, rebuild tgui

Timberpoes commented:
Okay, so. Here we go again.

As part of the code review I got to look into more detail at more areas of the code.

This time around how skillchips are stored.

It was undocumented and counter-intuitive that "deactivated" skillchips were just stored directly in the brain. It also felt a bit weird that activate skillchips were referenced in the list named "skillchips", instead of just "active_skillchips".

Now, skillchips are either `active = TRUE` or `active = FALSE` depending on if the last proc called was on_activate or on_deactivate. on_apply is now on_implant, while on_remove remains the same.

All code has been worked around this new convention. Skill chips can be activated or deactivated by calling on_activate and on_deactivate. Adding or removing skillchips from a brain (or a brain with skillchips from a body) will trigger on_implant and on_remove code. This means at all times, the skillchips list in the brain is **the** definitive list.

Timberpoes committed:
Minor logic fixes

Rohesie commented:
Just one review left from me. After that it looks good to me.
Up to @AnturK now.

AnturK commented on code/modules/mob/living/carbon/skillchip.dm:
```diff
@@ -0,0 +1,110 @@
+	if(dt < 0)
+		//Might deactivate more than necessary but not worth sorting this
+		while(dt < 0)
+			var/obj/item/skillchip/skillchip = brain.skillchips[length(brain.skillchips)]
```
Since this list now also contains inactive skillchips, you need to use something else here or you'll try to deactivate inactive ones.

AnturK reviewed (changes requested):
Just one oversight to fix and it's good to go.

TheSilverNuke commented:
Imagine emagging one of these skill chips.

Timberpoes committed:
Final Rohesie review

Timberpoes commented:
Standby for a day on this one.

Refactoring/rethinking another system (skillchip slots and how they're stored, referenced and held) whilst also upgrading tgui, then I should be complete. For another review. This time with added stylemistake.

This is to address one of AnturK's review notes. I'm aiming to make skillchips more streamlined in their use by allowing the skill station to activate and deactivate skillchips.

I'm also removing the slot_cost concept, which doesn't make sense in its current implementation, in favour of a complexity concept - A player can have as many chips as they like in their brain but can only activate them up to a max complexity. This functionality mirrors the current slot_cost, with the added bonus that the brain now stores the maximum complexity it can handle and will poll its owner body (if one exists) for any modifiers to that value.

This was to address concerns from people about switching jobs - You can now HOLD mutually exclusive chips (like two job chips at once) in your brain, but only ONE can be active at once. Activating a chip puts it on cooldown, so it can't be immediately deactivated.

Basically, a bunch more changes, but these should be the last system I'm not happy with, including pimping the interface.

Rohesie commented:
> This was to address concerns from people about switching jobs - You can now HOLD mutually exclusive chips (like two job chips at once) in your brain, but only ONE can be active at once. Activating a chip puts it on cooldown, so it can't be immediately deactivated.

Part of the reason for this was to diminish powergaming/hoarding. If you are adding this with a cooldown it should be no less than 10 minutes long.
Why do we want to allow for people to hoard skillchips anyway?

Timberpoes commented:
> Part of the reason for this was to diminish powergaming/hoarding. If you are adding this with a cooldown it should be no less than 10 minutes long.
> Why do we want to allow for people to hoard skillchips anyway?

The implementation was a bit strange. And was fiddly to use in practice.

I'm happy to implement an actual slot mechanic to complement complexity, but I want slots to be immutable. Complexity maximums can be flexible and changeable based on various player buffs. It'd add an additional layer of complexity and player choice too, compared to the existing system on master. The ability to have a few skillchips in your brain that you can activate and de-activate as necessary, with appropriate cooldowns. And it's a restriction of player choice, as thre's a hard limit to the number of chips in the brain regardless of complexity.

I'll start with the following system: 5 max chip slots, job and chameleon take up 2, everything else takes up 1, except the useless adaptor which takes up 0. This actually lets people hold more fluff 1-slot meme chips compared to live, but also allows people to have 2 job chips and 1 fluff chip which is what would happen on live too. 3 slot and higher chips become mutually exclusive with eachother just by virute of being too thicc.

And it'll just be a few number changes on some vars or datums to rebalance the whole system.

Timberpoes committed:
Robust code changes, improved slot/complexity system.

Timberpoes committed:
Rebuild tgui

Timberpoes commented:
The final big sweep. A new host of code changes including improvements across traitors and lings. Total rebuild of the interface.

Skill chips now have two variables - Complexity and Slot Use. Brains have both a complexity limit and a slot cost limit. This gives two natural balancing levers for skillchips that go beyond just arbitrary restrictions.

Slots represent the raw number of skill chips you can fit in your brain. It acts as both a skillchip storage for inactive skillchips and as a balancing mechanic; skillchips can be mutually exlcusive just based on how many slots they take up. This is controlled by a single var which can be changed in a couple of seconds. This var is held in the brain. Jobs chips and the chameleon chip take up 2 slots. There are a maximum of 5 slots. These are also easily changed as they're ordinary vars.

Complexity acts as a limit on which chips you can have active at the same time. Just because you have a bunch of chips in your brain, doesn't mean you can have all of them active at once. You can utilise multiple chips but may have to make choices and regularly return to the skill station to switch your active chips around if you want to utilise them fully. Complexity is the value affected by the biotech compatibility mutation. This var is also held in the brain. All chips currently have a complexity of 1. The maximum complexity is 3, increased to 4 with the biotech compatability mutation.

Chips don't go on cooldown when implanted or extracted. Instead, they go on cooldown when activated and deactivated, which by default is a 5 minute cooldown each way. A chip that is on cooldown cannot be extracted without applying a sharp instrument to a raw brain. The cooldown is set on a per-skillchip basis, currently nothing overwrites this and everything inherits 6 minutes from the base type.

The traitor chameleon chip has a new limitation; it cannot mimic chips with higher than its own slot cost. This adds an additional balancing lever to chameleon chips alongside blacklisting.

Changelings now inherit the skillchips at the exact state they were when they stung their target - Including activation state and cooldowns. Switching form inherits the chips. If you stung your target at a time they had all their chips deactivated with full cooldowns, tough luck!

In both of the above cases, these chips are non-recoverable and will be destroyed when extracted from a brain.

The interface has been improved, with additional visual feedback...
![image](https://user-images.githubusercontent.com/24975989/89285265-f5052a80-d647-11ea-8cdb-302eddfd51ee.png)

... and a very visual and icon-based interface with information tucked behind tooltips.

![image](https://user-images.githubusercontent.com/24975989/89285333-0fd79f00-d648-11ea-8728-6d198d1b578f.png)
![image](https://user-images.githubusercontent.com/24975989/89285364-1e25bb00-d648-11ea-80eb-176ee5a2308a.png)
![image](https://user-images.githubusercontent.com/24975989/89285396-2c73d700-d648-11ea-9254-4ab886765cec.png)

This should trigger a review from stylemistake where he yells at me for my formatting choices and asks me to pick better variable names.

Code has been tested locally as changes are made. A test run was completed with a full use of all functionality of the skill station, chameleon chip, skillchip autosurgeon and changeling, including changeling interaction with absorbing DNA from people with a chameleon chip. An extensive number of stack_traces and CRASHes can be found throughout the code to alert to abnormalities and unexpected behaviour so that issues can be addressed rapidly.

![image](https://user-images.githubusercontent.com/24975989/89286824-c5a3ed00-d64a-11ea-877f-4d7753fc5bfa.png)

Rohesie commented on code/modules/clothing/chameleon.dm:
```diff
@@ -672,6 +672,10 @@
 
+/obj/item/clothing/neck/chameleon/Destroy()
+	qdel(chameleon_action)
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/library/skill_learning/job_skillchips/traitor.dm:
```diff
@@ -0,0 +1,173 @@
+
+/obj/item/skillchip/chameleon/Destroy()
+	QDEL_NULL(chameleon_action)
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/library/skill_learning/job_skillchips/traitor.dm:
```diff
@@ -0,0 +1,173 @@
+
+/obj/item/skillchip/chameleon/set_metadata(var/list/metadata)
+	// Set base metadata first.
+	..()
```
```suggestion
	. = ..()
```

Timberpoes committed:
Code review

AnturK commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -249,13 +269,23 @@
+		skillchip.holding_brain = replacement_brain
+
+		// Directly add them to the skillchip list in the new brain.
+		LAZYADD(replacement_brain.skillchips, skillchip)
```
This might be bit dangerous if we add chips depending on specific brain subtype.

Timberpoes replied:
Sorted this out.

If the brain subtypes are different, a failsafe is engaged that attempts to deactivate all active skillchips. The player gets a feedback message about this event.

AnturK commented:
Are you done with changes ? If so this is good to merge.

Timberpoes commented:
Done with changes! This has come a long way from the simple chip I first intended to add, but I think this will serve as a robust framework for future job chips with multiple balancing levers.

Timberpoes committed:
Brain regeneration failsafe

Timberpoes commented:
One final, minor addition to cover AnturK's cases of brain subtypes in the future. If necessary a more bespoke check can be put in at some future point, but it's honestly not worth it for how rarely "regenerate organs and replace the existing brain" occurs in regular gameplay right now.

Timberpoes commented:
![image](https://user-images.githubusercontent.com/24975989/89445129-a5f3ee00-d74a-11ea-9ab0-0c812b08d497.png)

Timberpoes committed:
Lazylist cut fix
